compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala

```diff
@@ -46,14 +61,115 @@ class ExtractSemanticDB extends Phase:
+        ExtractSemanticDB.write(unit.source, extractor.occurrences.toList, extractor.symbolInfos.toList, extractor.synthetics.toList)
+    }
+    units
+  }
+
+  def run(using Context): Unit = unsupported("run")
+end ExtractSemanticDB
+
+object ExtractSemanticDB:
+  import java.nio.file.Path
+  import java.nio.file.Files
+  import java.nio.file.Paths
+
+  val phaseNamePrefix: String = "extractSemanticDB"
+  val description: String = "extract info into .semanticdb files"
+
+  enum PhaseMode:
+    case ExtractSemanticInfo
+    case AppendDiagnostics
+
+  class ExtractSemanticInfo extends ExtractSemanticDB(PhaseMode.ExtractSemanticInfo, "ExtractSemanticInfo")
+
+  class AppendDiagnostics extends ExtractSemanticDB(PhaseMode.AppendDiagnostics, "AppendDiagnostics")
+
+  private def semanticdbTarget(using Context): Option[Path] =
+    Option(ctx.settings.semanticdbTarget.value)
+      .filterNot(_.isEmpty)
+      .map(Paths.get(_))
+
+  private def outputDirectory(using Context): AbstractFile = ctx.settings.outputDir.value
+
+  private def absolutePath(path: Path): Path = path.toAbsolutePath.normalize
+
+  private def write(
+    source: SourceFile,
+    occurrences: List[SymbolOccurrence],
+    symbolInfos: List[SymbolInformation],
+    synthetics: List[Synthetic],
+  )(using Context): Unit =
+    val outpath = semanticdbPath(source)
+    Files.createDirectories(outpath.getParent())
+    val doc: TextDocument = TextDocument(
+      schema = Schema.SEMANTICDB4,
+      language = Language.SCALA,
+      uri = Tools.mkURIstring(Paths.get(relPath(source))),
+      text = "",
+      md5 = internal.MD5.compute(String(source.content)),
+      symbols = symbolInfos,
+      occurrences = occurrences,
+      synthetics = synthetics,
+    )
+    val docs = TextDocuments(List(doc))
+    val out = Files.newOutputStream(outpath)
+    try
+      val stream = internal.SemanticdbOutputStream.newInstance(out)
+      docs.writeTo(stream)
+      stream.flush()
+    finally
+      out.close()
+  end write
+
+  private def appendDiagnostics(
+    source: SourceFile,
+    diagnostics: Seq[Diagnostic]
+  )(using Context): Unit =
+    val path = semanticdbPath(source)
+    Using.Manager { use =>
+      val in = use(Files.newInputStream(path))
+      val sin = internal.SemanticdbInputStream.newInstance(in)
+      val docs = TextDocuments.parseFrom(sin)
+
+      val out = use(Files.newOutputStream(path))
+      val sout = internal.SemanticdbOutputStream.newInstance(out)
+      TextDocuments(docs.documents.map(_.withDiagnostics(diagnostics))).writeTo(sout)
+      sout.flush()
+    } match
+      case Failure(ex) => // failed somehow, should we say something?
+      case Success(_) => // success to update semanticdb, say nothing
+  end appendDiagnostics
+
+  private def relPath(source: SourceFile)(using ctx: Context) =
+    SourceFile.relativePath(source, ctx.settings.sourceroot.value)
+
+  private def semanticdbPath(source: SourceFile)(using ctx: Context) =
```

Comment:
I believe this does not need context
```suggestion
  private def semanticdbPath(source: SourceFile) =
```